### PR TITLE
Feature/data shares

### LINKED ISSUE #79: Create API endpoint for creating a data sharing url
This should create a unique token and store it with a name

### PR DESCRIPTION
**Pull request by Jakedavies**
closes #79

### CODE REVIEW

File: cg/src/domains/subjects/data-shares.controller.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const DataShareService = require('./data-shares.service');
+
+class DataShareController {
+  constructor(processorsService = null) {
```
Comment (sindelio) [inline cg/src/domains/subjects/data-shares.controller.js]: shouldn't this be a dataShareService instead of prcessorsService?

File: cg/src/utils/errors.js [hunk 2]
```diff
@@ -42,6 +42,7 @@ class NotFound extends Error {
 }
 
 const apiErrorHandler = (err, req, res, next) => {
+  console.log(err);
```
Comment (sindelio) [inline cg/src/utils/errors.js]: The intention is leaving this console.log here?

File: cg/test/subjects/data-shares.api.spec.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        method: 'GET',
+        headers: {
+          Authorization: `Bearer ${token}`
+        }
+      },
+      overrides
+    )
+  );
+}
+beforeAll(async () => {
```
Comment (sindelio) [inline cg/test/subjects/data-shares.api.spec.js]: Is it ok to have 2 beforeAll() and initResources()?

**Comment by sindelio** [comment]
What does it mean to create a data share in the system?

File: cg/test/subjects/data-shares.api.spec.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    method: 'POST',
+    body: { personalData: { name: 'test' } },
+    headers: {
+      Authorization: `Bearer ${token}`
+    }
+  });
+  return token;
+}
+
+function fetchWithAuthorization(path, token, overrides) {
```
Comment (sindelio) [inline cg/test/subjects/data-shares.api.spec.js]: What is the difference between this fetch and the other we were using?

File: cg/src/domains/subjects/data-shares.service.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    const data = {
+      subject_id: subjectId,
+      token,
+      name
+    };
+    await db('data_shares').insert(data);
+  }
+  async removeDataShare(dataShareId) {
+    const [dataShare] = await db('data_shares').where({ id: dataShareId });
+    if (!dataShare) throw new BadRequest('Data Share not found');
```
Comment (maksymilian-majer) [inline cg/src/domains/subjects/data-shares.service.js]: Shouldn't we rather throw `NotFound()`??

File: cg/src/domains/subjects/data-shares.service.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    await db('data_shares')
+      .delete()
```
[the remark this answers is not in the document]
Comment (maksymilian-majer) [reply cg/src/domains/subjects/data-shares.service.js]: The order does not matter and it could even be more SQL-like:
```
db.delete().from('data_shares').where({ id: dataShareId });
```
which reads like SQL:
```
DELETE FROM data_shares WHERE id = dataShareId
```

File: cg/test/subjects/data-shares.api.spec.js (hunk 4 above)
Comment (maksymilian-majer) [reply cg/test/subjects/data-shares.api.spec.js]: It is a wrapper that adds the `Authorization` header. @Jakedavies maybe we should extract it to `utils`, cause it's used all over the place?

File: cg/test/subjects/data-shares.api.spec.js (hunk 3 above)
Comment (maksymilian-majer) [reply cg/test/subjects/data-shares.api.spec.js]: I think this is a mistake

File: cg/test/subjects/data-shares.api.spec.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    await deployContract();
+  } catch (e) {
+    console.error(`Failed deploying contract ${e.toString()}`);
+  }
+  await initResources();
+});
+
+describe('Listing data-shares', () => {
+  it('should return the data-shares', async () => {
+    const subjectId = '673211';
```
Comment (maksymilian-majer) [inline cg/test/subjects/data-shares.api.spec.js]: since we have DB per test file we don't need to invent these random subjectIds anymore :)
that's cosmetics – I don't mean we should change it here

File: cg/test/subjects/data-shares.api.spec.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+beforeAll(async () => {
+  try {
+    await deployContract();
+  } catch (e) {
+    console.error(`Failed deploying contract ${e.toString()}`);
+  }
+  await initResources();
+});
+
+describe('Listing data-shares', () => {
```
Comment (maksymilian-majer) [inline cg/test/subjects/data-shares.api.spec.js]: I'd add a test case for listing data shares that returns an empty array – jut to be sure there's no HTTP 500

File: cg/test/subjects/data-shares.api.spec.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+  it('Should error if the share does not exist', async () => {
+    const subjectId = '8422291211';
+    const token = await createUser(subjectId);
+
+    const res = await fetchWithAuthorization(`/api/subject/data-shares/439101111/remove`, token, {
+      method: 'POST'
+    });
+
+    expect(res.status).toEqual(BadRequest.StatusCode);
```
Comment (maksymilian-majer) [inline cg/test/subjects/data-shares.api.spec.js]: As above – `BadRequest` vs `NotFound`?

File: cg/test/subjects/data-shares.api.spec.js (hunk 4 above)
Comment (Jakedavies) [reply cg/test/subjects/data-shares.api.spec.js]: yeah I think abstracting it makes sense, I wanted to play with it here to see if It saved any time.
